### PR TITLE
Better support for Kubernetes health probes

### LINKED ISSUE #18645: Podman kube play assumes port numbers in liveness probe
### Issue Description

Currently, when using Kubernetes liveness probes, podman assumes that Http/TCP port number is given in the probe, while you can use port name instead.  Podman should check if the livenessProbe.tcpSocket.port (and the equivalent HTTP one) is an string or a number, and if it is a string try to interpret it as a port name given in ports.name (which is checked as supported in podman-kube-play man page). 
 

### Steps to reproduce the issue

Steps to reproduce the issue
1. Use a Kybernetes YAML with livenessProbe.tcpSocket.port as a port name 
2. Run the pod using `podman kube play`, and then inspect the resulting pod 
3. The port in the given nc command is 0 (And the corresponding source code explicitly converts to Port to IntValue).

### Describe the results you received

0 port number used

### Describe the results you expected

The port number corresponding to the port name should be used instead.

### podman info output

```yaml
Tested with podman 4.5.0
```

### Podman in a container

No

### Privileged Or Rootless

Rootless

### Upstream Latest Release

Yes

### Additional environment details

Additional environment details

### Additional information

Additional information like issue happens only occasionally or issue happens with a particular architecture or on a particular setting

### PR DESCRIPTION
**Pull request by hedayat**
* Add support for using port names in Kubernetes probes
* Support TCPSocket probes without specifying any host names

Closes #18645

#### Does this PR introduce a user-facing change?

```release-note
From now on, you can also use containerPort names inside Kubernetes liveness probes to refer to the desired ports, which was not previously supported by podman.
```

### CODE REVIEW

**Comment by rhatdan** [comment]
@umohnani8 @ygalblum @baude PTAL

**Comment by rhatdan** [comment]
You need to add tests or add [NO NEW TESTS NEEDED] flag to the PR.

File: pkg/specgen/generate/kube/kube.go [hunk 1]
```diff
@@ -583,29 +586,53 @@ func probeToHealthConfig(probe *v1.Probe) (*manifest.Schema2HealthConfig, error)
...
 		path := "/"
 		if probeHandler.HTTPGet.Path != "" {
 			path = probeHandler.HTTPGet.Path
 		}
-		commandString = fmt.Sprintf("curl -f %s://%s:%d%s || %s", uriScheme, host, probeHandler.HTTPGet.Port.IntValue(), path, failureCmd)
+		portNum, err := getPortNumber(probeHandler.HTTPGet.Port, containerPorts)
+		if err != nil {
+			return nil, err
+		}
+		commandString = fmt.Sprintf("curl -f %s://%s:%d%s || %s", uriScheme, host, portNum, path, failureCmd)
```
Comment (TomSweeneyRedHat) [inline pkg/specgen/generate/kube/kube.go:600]: I know this is old code, but I don't believe curl is native to Mac.  If not, should we break this out into Linux, Mac, and Windows code streams?  Beyond which, I'm always leery about making a curl call from code, but that seems to be water under the bridge at this point.

Comment (hedayat) [reply pkg/specgen/generate/kube/kube.go:600]: Note that this command is being run inside the container, not the host. Although, this is also [not desirable](https://github.com/containers/podman/issues/18318). 

I guess the ideal solution would be either for podman itself to do the probes inside code, or to have a separate container (probably like or extending the pause container) running the probes.

**Comment by umohnani8** [comment]
This is a bit different than the issue I am looking at. What I am looking at is the probe fails with a curl command because the container doesn't have a shell or curl installed as the podman code currently runs the probe commands inside the container created. We need to figure out a way to handle http commands outside of the container to check for healthy status and will be a bigger rework.

I think this PR can go it as it is addressing another issue around ports with probes and there is no need to block it on the future rework.

Changes LGTM

**Comment by hedayat** [comment]
> This is a bit different than the issue I am looking at. What I am looking at is the probe fails with a curl command because the container doesn't have a shell or curl installed as the podman code currently runs the probe commands inside the container created. We need to figure out a way to handle http commands outside of the container to check for healthy status and will be a bigger rework.

That's #18318, and its also the case for TCP probes which is also  done by running `nc` inside the container. And there will be gRPC probes when implemented in podman :)


> Changes LGTM

Thanks a lot for the review!

File: pkg/specgen/generate/kube/kube.go [hunk 2]
```diff
@@ -569,6 +569,7 @@ func probeToHealthConfig(probe *v1.Probe, containerPorts []v1.ContainerPort) (*m
 	var commandString string
 	failureCmd := "exit 1"
 	probeHandler := probe.Handler
+	host := "localhost" // Kubernetes default is host IP, but with Podman currently we run inside the container
```
Comment (vrothberg) [inline pkg/specgen/generate/kube/kube.go:572]: Any chance we can add a test for it?

**Review by vrothberg — COMMENTED** [review]
I´d love to see a test or two being added (or extended) in https://github.com/containers/podman/blob/main/test/e2e/play_kube_test.go.

**Comment by ygalblum** [comment]
LGTM for the code. But, I agree with @vrothberg about the need for tests

**Comment by rhatdan** [comment]
/lgtm
I don't think this is a new feature ,so it should be back ported.

**Pull request merged**